Fix bug in handling delete key event for android

alimahdiyar opened the pull request:
When typing RTL and LTR characters together in TextField, deleting a single character might cause it's previous characters from the same direction to be deleted.
Also, you can't delete a new line when typing RTL text.

This PR is for fixing these problems.

justinmc commented on shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:
```diff
@@ -241,21 +241,39 @@ public boolean sendKeyEvent(KeyEvent event) {
+
+          Boolean selectedCharIsEmoji = false;
+          if (selStart > 1) {
+            final String emo_regex =
```
Should be camel case, like `emojiRegex`.

justinmc commented on shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:
```diff
@@ -241,21 +241,39 @@ public boolean sendKeyEvent(KeyEvent event) {
+          Boolean selectedCharIsEmoji = false;
+          if (selStart > 1) {
+            final String emo_regex =
+                "([\\u20a0-\\u32ff\\ud83c\\udc00-\\ud83d\\udeff\\udbb9\\udce5-\\udbb9\\udcee])";
```
Where did this regex come from?  Are you sure this will match exactly the characters that you want it to match?

alimahdiyar replied:
Thanks for your review!
I'm now doing research on the regex that matches all needed characters.

justinmc commented on shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:
```diff
@@ -241,21 +241,39 @@ public boolean sendKeyEvent(KeyEvent event) {
+          if (selectedCharIsEmoji) {
+            boolean isRtl =
+                mLayout.isRtlCharAt(mLayout.getLineStart(mLayout.getLineForOffset(selStart)));
+            Log.i("isRtl", Boolean.toString(isRtl));
```
I think you can delete this Log.

alimahdiyar commented:
There was also another problem in the main code
`boolean isRtl = mLayout.isRtlCharAt(mLayout.getLineForOffset(selStart));`
should be:
`boolean isRtl = mLayout.isRtlCharAt(mLayout.getLineStart(mLayout.getLineForOffset(selStart)))`
for getting the first character in the line.

alimahdiyar commented:
I wrote some test for the problems I mentioned above.
I also changed my code and simply used android QwertyKeyListener which solved all the problems.
Please review my PR again.
Thanks.

justinmc commented:
@alimahdiyar Looks like there is a conflict, could you merge master?

alimahdiyar commented:
> @alimahdiyar Looks like there is a conflict, could you merge master?

Done

alimahdiyar commented:
> @alimahdiyar Looks like there is a conflict, could you merge master?

@justinmc Master is merged and all android checks are successful.
Tried to manually re-run some checks because they got stuck and they're marked canceled.
However, there is no reason for windows build check to fail.

alimahdiyar commented:
> @alimahdiyar Looks like there is a conflict, could you merge master?

@justinmc Master is merged and all checks are successfully passed.

alimahdiyar commented:
This is also a better fix for #9734 that handles emojis @GaryQian

GaryQian commented:
Thanks for the changes, give me a moment to verify this is valid for all the different kinds of input.

GaryQian commented on shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:
```diff
@@ -230,35 +232,12 @@ public boolean sendKeyEvent(KeyEvent event) {
-            // back to a simple way of deletion, although this a hack and
-            // will not handle emojis.
-            Selection.setSelection(mEditable, selStart, selStart - 1);
+          QwertyKeyListener qwertyKeyListener = new QwertyKeyListener(Capitalize.NONE, false);
```
This object can be cached instead of re-created every time, though as usual it is a trade-off between speed and memory usage. Since keystrokes are manual rare events, caching this is optional. Your choice.

alimahdiyar replied:
I'm now using getInstance() which handles caching.

GaryQian commented:
It looks like there is a listener called `TextKeyListener` which may be a more generic (non-QWERTY specific) class to use.

If `TextKeyListener` does the same job as `QwertyKeyListener`, I would prefer if we used `TextKeyListener` instead, as it should be less likely to run into internationalization issues.

GaryQian commented:
According to the API docs for `QwertyKeyListener`, `QwertyKeyListener` should not be instantiated directly. Instead, a `TextKeyListener` should be used which will multiplex and instantiate multiple keyboard-specific listeners.

alimahdiyar commented:
Thank you for reviewing my PR.

I made the needed changes to my code and I'm now using TextKeyListener class.

GaryQian commented on shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:
```diff
@@ -230,35 +231,11 @@ public boolean sendKeyEvent(KeyEvent event) {
-            // back to a simple way of deletion, although this a hack and
-            // will not handle emojis.
-            Selection.setSelection(mEditable, selStart, selStart - 1);
+          if (TextKeyListener.getInstance().onKeyDown(null, mEditable, event.getKeyCode(), event)) {
```
I tested with a few keyboard, and it looks like this isn't quite working with extended grapheme cluster emojis, such as flag emojis: Type a few flag emojis, and try to delete them from the middle (move caret to middle, then delete) On a few of my devices, I am seeing the flag emoji only partially delete and turn into its components.

GaryQian replied:
It looks like the same happens with the Qwerty version of this.

justinmc replied:
Does it still happen without this PR?  On master I've noticed I can turn a grapheme cluster emoji into its components using the arrow keys to put the cursor in the middle.

alimahdiyar replied:
I think the problem is another bug in moving cursor around. When touching TextField on a real device I got no bugs. But when dragging it (sometimes) or moving with keyboard arrows, cursor is not placed correctly.
When the cursor place was correct, my code worked.
Did you experience the problem on real devices and touch interfaces?
If yes, give me more information, please

GaryQian replied:
Yes, this was all on real devices with touch.

Reproed on:
SM-T380 (Galaxy Tab)
Pixel 3
Samsung Galaxy Note 9
Huawei Nova 4

Here is a screenshot:
![flutter_01](https://user-images.githubusercontent.com/1887398/78728191-8b6c4e00-78eb-11ea-803e-3a34ff14c3fb.png)

Directions: Create app with basic TextField widget
Type any flag emojis
Tap in the middle, caret will appear correctly between flags.
Press backspace. The flag emoji to be deleted will break apart, with the letter emoji and a question mark representing the zero width joiner.

The reason is that the key listener is deleting by codepoint. This is fine for simple emojis, but will break more complex emojis such as family emojis which depend on many component emojis linked with "zero width joiner" characters in between. The goal of the expandLeft and expandRight is that those android methods would expand by glyph, anutomatically handling any codepoints that would be encountered. The bug is caused by RTL scenarios mixing with LTR characters, which may appear on the visually wrong side to be encountered by the expand, which expands visually.

alimahdiyar replied:
I think caret place is visually correct. I'm currently working on a better solution or at least a hotfix for RTL scenarios

alimahdiyar replied:
Are you sure this bug doesn't happen without this PR?

alimahdiyar replied:
The reason I say the problem might be with caret place is that when the caret was in the end or when I placed it correctly with arrow keys, the code could delete a complete glyph.

alimahdiyar commented:
I worked on the code and updated my PR and mixed my solution with the old one.
These are things that are fixed by this PR:
1. Fix detecting RTL line logic
2. Fix newline not being deleted after an RTL line.
3. Fix the bug that multiple characters were deleted when mixing RTL and LTR characters.
4. Handle simple emojis on Chinese devices having problem with extendRight or extendLeft.

P.S: extendLeft didn't work correctly on tests. I'm not sure but I think it needed a physical device. Anyway, I deleted the tests but confirmed everything works as good as possible on real devices.

Please review again. @GaryQian 
Thanks for your attention.

alimahdiyar commented:
@GaryQian Can you please review again?
I'm working on some other issues in TextAlign and Direction too and some people might be waiting for them to use their apps for production.

GaryQian commented:
So I tested with flag/complex emojis and it looks like the new version still does not handle them properly. Let me see if I can wiggle it around to handle the emojis properly. I agree with the RTL handling changes, but I would not want to regress the emoji backspace behavior. Simple emojis should be no problem as they are represented by single codepoints.

alimahdiyar commented:
> So I tested with flag/complex emojis and it looks like the new version still does not handle them properly. Let me see if I can wiggle it around to handle the emojis properly. I agree with the RTL handling changes, but I would not want to regress the emoji backspace behavior. Simple emojis should be no problem as they are represented by single codepoints.

When not mixing RTL and LTR text and writing simple one direction text, my code is using old extendRight and extendLeft functions. So if the bug exists in this scenarios the bug origin is not my code.

If the problem is when you mix RTL and LTR text (which is handled by the new code), it has the benefit of fixing more problems.

Thanks for your review again and working on the problem
I love flutter and I'm happy if I could make it grow better.

GaryQian commented:
Hang in there, I am experiencing a strange issue where complex emojis are not being handled even with a clean, unmodified engine build. It seems that the debug local engine is somehow causing improper emoji handling?

alimahdiyar commented:
I didn't have devices that you reported the issue and didn't experience the issue so I couldn't test myself. But as I said I think the problem is with bringing the caret in the wrong offset when placing.
I have a suggestion.
Log the caret offset when you tap the TextField and reassign the caret place to see if the offset is between glyphs or not.

GaryQian commented:
It's not that, your code might have been correct. When I run everything in release mode, it seems to start working. Still trying to isolate why this is happening. I haven't seen debug vs release differences like this before.

GaryQian commented:
Ok, so I have confirmed that the version using just the TextKeyListener directly is producing correct results when built as a release engine. If you roll it back to that version, I can approve it. I'm not sure how there can be differences in the debug vs release though, the only possible causes are either a timing issue (the keyboard API can get race-y due to the jumps between dart, engine, and embedder) or a dart compiler bug where there is some tiny difference between the two builds.

The GetGlyphPositionAtCoordinate() method in LibTxt is simply not being called, resulting in the incorrect caret positioning as you pointed out. Since this bug is unrelated to your fix, we can handle them separately.

alimahdiyar commented:
Done
Thanks again

alimahdiyar commented:
@GaryQian Thanks.
I'm not familiar with this repository's merge process. Your approval means this will be merged automatically?

GaryQian commented:
I will take care of merging it :). I'll let you know if anything needs to be fixed to pass presubmit.

GaryQian commented:
@alimahdiyar Can you rebase to upstream master? There was a change made recently that is needed for the `Mac iOS Engine` test to pass.

`git pull upstream master --rebase` should do it. Thanks!

alimahdiyar commented:
@GaryQian Done :)

GaryQian commented:
Landing manually, Fuchsia LUCI failure seems to be a flake.
